File: packages/domain/src/merkle.ts
import { createHash } from "node:crypto";
import type { Allocation } from "./distribution.js";

/**
 * ODP Merkle allocation format — **MERKLE-WIRE-FORMAT = PROVISIONAL**.
 * Not frozen until the TypeScript and Rust/Solana implementations produce
 * identical cross-language test vectors (leaf bytes, leaf hash, tree root,
 * proof, verification result) in P0-4 / G4. Do NOT swap SHA-256 for another
 * hash before measured CU data from the on-chain program justifies it.
 *
 *   leaf   = sha256(UTF8(distribution_id + "\n" + wallet + "\n" + amount))
 *   node   = sha256(concat(sort(child_a, child_b)))
 * Leaves are sorted by Buffer.compare before pairing (input order does not
 * affect the root); sibling sorting makes proofs direction-free; an odd
 * trailing node is promoted unchanged. Duplicate leaves are rejected.
 * distribution_id inside the leaf blocks cross-distribution proof replay.
 */
export function allocationLeaf(a: Pick<Allocation, "distribution_id" | "wallet" | "amount">): Buffer {
  return createHash("sha256")
    .update(`${a.distribution_id}\n${a.wallet}\n${a.amount}`, "utf8")
    .digest();
}

function hashPair(a: Buffer, b: Buffer): Buffer {
  const [x, y] = Buffer.compare(a, b) <= 0 ? [a, b] : [b, a];
  return createHash("sha256").update(Buffer.concat([x, y])).digest();
}

export interface MerkleTree {
  root: Buffer;
  /** Leaves in tree (sorted) order. */
  leaves: Buffer[];
  /** levels[0] = sorted leaves; last level = [root]. */
  levels: Buffer[][];
}

export function buildMerkleTree(leaves: Buffer[]): MerkleTree {
  if (leaves.length === 0) throw new Error("merkle tree requires at least one leaf");

  const sorted = [...leaves].sort(Buffer.compare);
  for (let i = 1; i < sorted.length; i++) {
    if (sorted[i - 1]!.equals(sorted[i]!)) {
      throw new Error("duplicate merkle leaf (same wallet+amount appears twice)");
    }
  }

  const levels: Buffer[][] = [sorted];
  let level = sorted;
  while (level.length > 1) {
    const next: Buffer[] = [];
    for (let i = 0; i < level.length; i += 2) {
      const a = level[i]!;
      const b = i + 1 < level.length ? level[i + 1]! : null;
      next.push(b === null ? a : hashPair(a, b));
    }
    levels.push(next);
    level = next;
  }
  return { root: level[0]!, leaves: sorted, levels };
}

/** Proof for one leaf, identified by buffer equality against the sorted leaves. */
export function merkleProof(tree: MerkleTree, leaf: Buffer): Buffer[] {
  let index = tree.leaves.findIndex((l) => l.equals(leaf));
  if (index === -1) throw new Error("leaf not present in tree");

  const proof: Buffer[] = [];
  for (const level of tree.levels.slice(0, -1)) {
    const siblingIndex = index % 2 === 0 ? index + 1 : index - 1;
    const sibling = siblingIndex < level.length ? level[siblingIndex]! : null;
    // Odd trailing node has no sibling — nothing to fold for this level.
    if (sibling !== null) proof.push(sibling);
    index = Math.floor(index / 2);
  }
  return proof;
}

/** Direction-free verification: fold leaf through the proof via sorted pairing. */
export function verifyMerkleProof(leaf: Buffer, proof: Buffer[], root: Buffer): boolean {
  let node = leaf;
  for (const p of proof) node = hashPair(node, p);
  return node.equals(root);
}

File: programs/merkle-vector/src/merkle.rs
//! Rust mirror of the ODP merkle wire format (must stay byte-identical to
//! packages/domain/src/merkle.ts):
//!   leaf = sha256(utf8(distribution_id + "\n" + wallet + "\n" + amount))
//!   node = sha256(concat(sort(child_a, child_b)))
//! Leaves sorted before pairing; sibling sorting makes proofs
//! direction-free; odd trailing node promoted unchanged.

use crate::sha256::sha256;

pub fn leaf(distribution_id: &str, wallet: &str, amount: &str) -> [u8; 32] {
    let preimage = format!("{}\n{}\n{}", distribution_id, wallet, amount);
    sha256(preimage.as_bytes())
}

fn hash_pair(a: &[u8], b: &[u8]) -> [u8; 32] {
    let (x, y) = if a <= b { (a, b) } else { (b, a) };
    let mut buf = Vec::with_capacity(x.len() + y.len());
    buf.extend_from_slice(x);
    buf.extend_from_slice(y);
    sha256(&buf)
}

pub struct Tree {
    pub root: [u8; 32],
    pub leaves: Vec<[u8; 32]>,
    pub levels: Vec<Vec<[u8; 32]>>,
}

pub fn build(leaves: &[[u8; 32]]) -> Tree {
    assert!(!leaves.is_empty(), "merkle tree requires at least one leaf");
    let mut sorted = leaves.to_vec();
    sorted.sort();
    for i in 1..sorted.len() {
        assert!(sorted[i - 1] != sorted[i], "duplicate merkle leaf");
    }

    let mut levels: Vec<Vec<[u8; 32]>> = vec![sorted.clone()];
    while levels.last().unwrap().len() > 1 {
        let level = levels.last().unwrap();
        let mut next: Vec<[u8; 32]> = Vec::with_capacity((level.len() + 1) / 2);
        let mut i = 0;
        while i < level.len() {
            if i + 1 < level.len() {
                next.push(hash_pair(&level[i], &level[i + 1]));
            } else {
                next.push(level[i]);
            }
            i += 2;
        }
        levels.push(next);
    }

    Tree {
        root: levels.last().unwrap()[0],
        leaves: sorted,
        levels,
    }
}

pub fn proof(tree: &Tree, leaf_bytes: &[u8; 32]) -> Vec<[u8; 32]> {
    let mut index = tree
        .leaves
        .iter()
        .position(|l| l == leaf_bytes)
        .expect("leaf not present in tree");
    let mut out: Vec<[u8; 32]> = Vec::new();
    for level in tree.levels.iter().take(tree.levels.len() - 1) {
        let sibling_index = if index % 2 == 0 { index + 1 } else { index - 1 };
        if sibling_index < level.len() {
            out.push(level[sibling_index]);
        }
        index /= 2;
    }
    out
}

pub fn verify(leaf_bytes: &[u8; 32], proof: &[[u8; 32]], root: &[u8; 32]) -> bool {
    let mut node = *leaf_bytes;
    for p in proof {
        node = hash_pair(&node, p);
    }
    &node == root
}
